### PR TITLE
Adds functional tests for self service wallets.

### PR DESCRIPTION
**Pull request by mango-habanero**

### CODE REVIEW

File: test/functional/app/api/test_ussd_api.py [hunk 1]
```diff
@@ -127,11 +141,66 @@ def test_golden_path_send_token(mocker, test_client,
...
-        pin_hash=User.salt_hash_secret('0000'), default_organisation=org)
+                         pin_hash=User.salt_hash_secret('0000'), default_organisation=org)
 
-    resp = req("", test_client, sender, '*42*666#')
+    resp = req("", test_client, sender.phone, '*42*666#')
     assert 'END Bonyeza {} kutumia mtandao'.format(valid_service_code) in resp
+
+
+@pytest.fixture(scope='module')
+def create_temporary_user(test_client, init_database, create_organisation):
```
Comment (nolash) [inline test/functional/app/api/test_ussd_api.py]: This is a fixture so it should be in a `conftest.py` file. Judging from context it looks like it should be in the root one.

File: test/functional/app/api/test_ussd_api.py [hunk 2]
```diff
@@ -127,11 +141,66 @@ def test_golden_path_send_token(mocker, test_client,
...
     sender = UserFactory(preferred_language="en", phone=make_kenyan_phone(phone()), first_name="Bob", last_name="Foo",
-        pin_hash=User.salt_hash_secret('0000'), default_organisation=org)
+                         pin_hash=User.salt_hash_secret('0000'), default_organisation=org)
 
-    resp = req("", test_client, sender, '*42*666#')
+    resp = req("", test_client, sender.phone, '*42*666#')
     assert 'END Bonyeza {} kutumia mtandao'.format(valid_service_code) in resp
+
+
+@pytest.fixture(scope='module')
```
Comment (nolash) [inline test/functional/app/api/test_ussd_api.py]: scope function because you would be manipulating the user from scratch each test, no?

File: test/functional/app/api/test_ussd_api.py [hunk 3]
```diff
@@ -127,11 +141,66 @@ def test_golden_path_send_token(mocker, test_client,
...
+def create_temporary_user(test_client, init_database, create_organisation):
+    # create organisation
+    organisation = create_organisation
+    organisation.external_auth_password = config.EXTERNAL_AUTH_PASSWORD
+
+    # set active organisation
+    g.active_organisation = organisation
+
+    # create user without a transfer account
+    user = create_user_without_transfer_account(unregistered_user_phone)
```
Comment (nolash) [inline test/functional/app/api/test_ussd_api.py]: I see we have no test for the "attach transfer account" hack from last time. Could we have one please?

File: test/functional/app/api/test_ussd_api.py [hunk 4]
```diff
@@ -127,11 +141,66 @@ def test_golden_path_send_token(mocker, test_client,
 
 
 def test_invalid_service_code(mocker, test_client,
-                                init_database, initialised_blockchain_network, init_seed):
-
+                              init_database, initialised_blockchain_network, init_seed):
     org = OrganisationFactory()
     sender = UserFactory(preferred_language="en", phone=make_kenyan_phone(phone()), first_name="Bob", last_name="Foo",
-        pin_hash=User.salt_hash_secret('0000'), default_organisation=org)
+                         pin_hash=User.salt_hash_secret('0000'), default_organisation=org)
```
Comment (nolash) [inline test/functional/app/api/test_ussd_api.py:159]: why the extra indent? please either single indent or each arg on separate lines.

File: test/functional/app/api/test_ussd_api.py [hunk 5]
```diff
@@ -127,11 +141,66 @@ def test_golden_path_send_token(mocker, test_client,
...
+                               create_temporary_user,
+                               create_organisation):
+
+    # create organisation
+    organisation = create_organisation
+    organisation.external_auth_password = config.EXTERNAL_AUTH_PASSWORD
+
+    # external auth password matching model definition of the same value.
+    # org.external_auth_username = 'admin_'+(org.name or '').lower().replace(' ', '_') /961ab9adc300_.py
+    external_auth_username = 'admin_' + (organisation.name or '').lower().replace(' ', '_')
```
Comment (nolash) [inline test/functional/app/api/test_ussd_api.py:175]: Let's re-use the same code that creates it in the app as a function.

File: test/functional/app/api/test_ussd_api.py [hunk 6]
```diff
@@ -127,11 +141,66 @@ def test_golden_path_send_token(mocker, test_client,
...
+
+    user = create_temporary_user
+
+    resp = req("1", test_client, user.phone, auth_username=external_auth_username)
+    assert "CON Please enter a PIN" in resp
+
+    resp = req("0000", test_client, user.phone, auth_username=external_auth_username)
+    assert "CON Enter your PIN again" in resp
+
+    resp = req("0000", test_client, user.phone, auth_username=external_auth_username)
```
Comment (nolash) [inline test/functional/app/api/test_ussd_api.py:191]: Perhaps we should have a separate test for wrong pin, and retry with correct pins. This specific case made trouble at some point.

File: test/functional/app/api/test_ussd_api.py [hunk 7]
```diff
@@ -2,6 +2,8 @@
 from faker.providers import phone_number
 from faker import Faker
 from functools import partial
+from contextlib import contextmanager
```
Comment (nolash) [inline test/functional/app/api/test_ussd_api.py]: import not used?

File: test/functional/app/api/test_ussd_api.py [hunk 8]
```diff
@@ -2,6 +2,8 @@
...
+from flask import appcontext_pushed, g
```
Comment (nolash) [inline test/functional/app/api/test_ussd_api.py]: `appcontext_pushed` not used?

File: test/functional/app/api/test_ussd_api.py [hunk 9]
```diff
@@ -127,11 +141,66 @@ def test_golden_path_send_token(mocker, test_client,
...
+
+    resp = req("0000", test_client, user.phone, auth_username=external_auth_username)
+    assert "END Your account is being created." in resp
+    assert user.first_name == 'Unknown first name'
+    assert user.last_name == 'Unknown last name'
+    bio = next(filter(lambda x: x.name == 'bio', user.custom_attributes), None)
+    assert bio.value.strip('"') == 'Unknown business'
+    gender = next(filter(lambda x: x.name == 'gender', user.custom_attributes), None)
+    assert gender.value.strip('"') == 'Unknown gender'
+    assert user.location == 'Unknown location'
```
Comment (nolash) [inline test/functional/app/api/test_ussd_api.py:199]: Is it worth checking that the consecutive ussd session gives the usual menu?

File: test/functional/app/api/test_ussd_api.py [hunk 10]
```diff
@@ -127,11 +141,66 @@ def test_golden_path_send_token(mocker, test_client,
...
+
+    resp = req("0000", test_client, user.phone, auth_username=external_auth_username)
+    assert "CON Enter your PIN again" in resp
+
+    resp = req("0000", test_client, user.phone, auth_username=external_auth_username)
+    assert "END Your account is being created." in resp
+    assert user.first_name == 'Unknown first name'
+    assert user.last_name == 'Unknown last name'
+    bio = next(filter(lambda x: x.name == 'bio', user.custom_attributes), None)
+    assert bio.value.strip('"') == 'Unknown business'
```
Comment (nolash) [inline test/functional/app/api/test_ussd_api.py]: hmm suspicious that those values are returned encapsulated by  `"` ?

**Review by nolash — CHANGES_REQUESTED** [review]

File: test/functional/app/api/test_ussd_api.py (hunk 5 above)
Comment (mango-habanero) [reply test/functional/app/api/test_ussd_api.py:175]: The code that creates in the app is a migration.
It is also really annoying, the discrepancy in what the name will be in the DB and what the config file is looking for

File: test/functional/app/api/test_ussd_api.py (hunk 6 above)
Comment (mango-habanero) [reply test/functional/app/api/test_ussd_api.py:191]: Noted, this will be added

File: test/functional/app/api/test_ussd_api.py (hunk 7 above)
Comment (mango-habanero) [reply test/functional/app/api/test_ussd_api.py]: Will clean up

File: test/functional/app/api/test_ussd_api.py (hunk 2 above)
Comment (mango-habanero) [reply test/functional/app/api/test_ussd_api.py]: Yes

File: test/functional/app/api/test_ussd_api.py (hunk 1 above)
Comment (mango-habanero) [reply test/functional/app/api/test_ussd_api.py]: I grappled with putting it there but then its only used once in this very specific context ? Should such fixtures also be in conftest ?

File: test/functional/app/api/test_ussd_api.py (hunk 10 above)
Comment (mango-habanero) [reply test/functional/app/api/test_ussd_api.py]: It would seem so. I'll confirm this though

File: test/functional/app/api/test_ussd_api.py (hunk 8 above)
Comment (mango-habanero) [reply test/functional/app/api/test_ussd_api.py]: Clean up as well

File: test/functional/app/api/test_ussd_api.py (hunk 1 above)
Comment (nolash) [reply test/functional/app/api/test_ussd_api.py]: I believe it should be in a `conftest.py` file. But that could be a separate file in this directory, too. If I recall correctly parent dirs are checked in order for that filename to include all fixtures.

File: test/functional/app/api/test_ussd_api.py (hunk 5 above)
Comment (mango-habanero) [reply test/functional/app/api/test_ussd_api.py:175]: The code that creates it in the app is a migration, I don't think we can reuse it.

Comment (nolash) [reply test/functional/app/api/test_ussd_api.py:175]: Ok let's not spend time on this. If it's the same as in the migration that's fine then.

File: test/functional/app/api/test_ussd_api.py (hunk 9 above)
Comment (nolash) [reply test/functional/app/api/test_ussd_api.py:199]: @PhilipWafula no comment on this?

File: test/unit/app/utils/test_utils_user.py [hunk 11]
```diff
@@ -18,7 +19,14 @@
 #     assert create_transfer_account_user.one_time_code is not None
 #     assert create_transfer_account_user.transfer_account is not None
 from helpers.factories import UserFactory, OrganisationFactory, TokenFactory, TransferAccountFactory, fake
-from server.utils.user import transfer_usages_for_user, send_onboarding_sms_messages, admin_reset_user_pin, proccess_create_or_modify_user_request
+from server.models.user import RegistrationMethodEnum
```
Comment (nolash) [inline test/unit/app/utils/test_utils_user.py:22]: Oh please on one line, rather.

In general when we need so many symbols from one package that suggests we should have some sort of helper instead perhaps. We can look at that when and if we ever get around to cleaning up.

File: test/unit/app/utils/test_utils_user.py [hunk 12]
```diff
@@ -140,3 +148,29 @@ def test_send_welcome_sms(mocker, test_client, init_database, mock_sms_apis,
     messages = mock_sms_apis
     assert messages == [{'phone': '+61123456789', 'message': expected_welcome},
                         {'phone': '+61123456789', 'message': expected_terms},]
+
+def test_attach_transfer_account_to_user(test_client, init_database, create_organisation):
+    with db.session.no_autoflush:
```
Comment (nolash) [inline test/unit/app/utils/test_utils_user.py:149]: why is this needed?

File: test/unit/app/utils/test_utils_user.py [hunk 13]
```diff
@@ -140,3 +148,29 @@ def test_send_welcome_sms(mocker, test_client, init_database, mock_sms_apis,
...
+                           registration_method=RegistrationMethodEnum.USSD_SIGNUP)
+
+        user.add_user_to_organisation(organisation, False)
+
+        assert user.primary_blockchain_address is None
+
+        user_with_transfer_account = attach_transfer_account_to_user(user)
+
+        assert user_with_transfer_account.primary_blockchain_address is not None
+        bio = next(filter(lambda x: x.name == 'bio', user_with_transfer_account.custom_attributes), None)
```
Comment (nolash) [inline test/unit/app/utils/test_utils_user.py]: are the custom attributes linked to the transfer account? if not, let's remove the checks that aren't in context.

**Review by nolash — CHANGES_REQUESTED** [review]

File: test/functional/app/api/test_ussd_api.py [hunk 14]
```diff
@@ -2,81 +2,106 @@
...
-from server.utils.user import default_transfer_account
+from server.utils.user import default_transfer_account, create_user_without_transfer_account
 
 fake = Faker()
 fake.add_provider(phone_number)
 phone = partial(fake.msisdn)
+unregistered_user_phone = make_kenyan_phone(phone())
+
 
 @pytest.fixture(scope='module')
```
Comment (nolash) [inline test/functional/app/api/test_ussd_api.py]: this fixture should probably also be somewhere else?

File: test/unit/app/utils/test_utils_user.py (hunk 12 above)
Comment (mango-habanero) [reply test/unit/app/utils/test_utils_user.py:149]: Tests that the transfer account is actually attached to a self service user once the post request to process the transfer account creation.
Feels like its implicitly tested in the functional tests for the self service flow now that you've brought it to my attentions.
Its components could largely just be incorporated to the test in the func tests.

Comment (nolash) [reply test/unit/app/utils/test_utils_user.py:149]: sorry I meant the autoflush - and we discucsed already :)

File: test/functional/app/api/test_ussd_api.py [hunk 15]
```diff
@@ -236,3 +235,64 @@ def test_ussd_self_signup_wrong_pin_entry(test_client,
...
+
+    # define different session id
+    other_session_id = 'ATUid_05af06225e6163ec2dc9dc9cf8bc97aa000'
+
+    # create organisation
+    organisation = create_organisation
+    organisation.external_auth_password = config.EXTERNAL_AUTH_PASSWORD
+
+    # external auth password matching model definition of the same value.
+    # org.external_auth_username = 'admin_'+(org.name or '').lower().replace(' ', '_') /961ab9adc300_.py
```
Comment (nolash) [inline test/functional/app/api/test_ussd_api.py]: please remove commented code

**Review by nolash — COMMENTED** [review]
@PhilipWafula thank you for your patience and for amending the code. Please see just one more small thing about commented code and then I will approve.

File: test/functional/app/api/test_ussd_api.py (hunk 15 above)
Comment (mango-habanero) [reply test/functional/app/api/test_ussd_api.py]: The comment isn't part of the code in anyway, it was there to provide context for why that `external_auth_name` was structured like that, in case we or someone else ever comes back to fix this, I feel the context would be helpful.
But then again it could be a less confusing comment so let me just make it clearer real quick.

**Review by nolash — APPROVED** [review]
